### DISCUSSION TITLE
[Share] Auto-tag when keyword detected in title

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-actions-tags#119] [Action Scripts]
## Description

Auto-tag when keyword detected in title. See #77 #85.

## Action Settings

Event: Create Item
Type: Script
Data:
```js
// Auto-tag when keyword detected in title
// @author windingwind
// @link https://github.com/windingwind/zotero-actions-tags/discussions/119
// @usage Replace the `KEYWORD` and `TAG_NAME`.
// @see https://github.com/windingwind/zotero-actions-tags/issues/77
// @see https://github.com/windingwind/zotero-actions-tags/issues/85

if (!item) {
	return "[Conditional Tag] item is empty";
}

const KEYWORD = "REPLACE WITH YOUR KEYWORD";
const TAG_NAME = "REPLACE WITH YOUR TAG"

if (item.getField("title").includes(KEYWORD)) {
	item.addTag(TAG_NAME, 1);
	return `[Conditional Tag] item ${item.getField("title")} with keyword ${KEYWORD} is  tagged ${TAG_NAME}`;
}

return `[Conditional Tag] item ${item.getField("title")} does not contain keyword ${KEYWORD}`
```

**Comment by thy960112** [comment]
Thanks, this feature is very useful. How can I use multiple keywords and tags in only one script? More specifically, if the title contains keywords A or B, add tags named C and D.
I have tried separating multiple keywords with commas like `A, B`, and the expected result is not to check each keyword individually, but to check together and determine that the title does not contain `A, B`.

**Reply by windingwind** [reply]
```js
// Auto-tag when keywords detected in title
// @author windingwind
// @link https://github.com/windingwind/zotero-actions-tags/discussions/119
// @usage Replace the `KEYWORDS` and `TAG_NAMES`.
// @see https://github.com/windingwind/zotero-actions-tags/issues/77
// @see https://github.com/windingwind/zotero-actions-tags/issues/85

if (!item) {
	return "[Conditional Tag] item is empty";
}

const KEYWORDS = ["KWD1", "KWD2", "KWD3"];
const TAG_NAMES = ["TAG1", "TAG2"];

if (KEYWORDS.some(kwd => item.getField("title").includes(kwd))) {
	TAG_NAMES.forEach(tag => item.addTag(tag, 1));
	return `[Conditional Tag] item ${item.getField("title")} with keyword ${KEYWORDS} is  tagged ${TAG_NAMES}`;
}

return `[Conditional Tag] item ${item.getField("title")} does not contain keyword ${KEYWORDS}`;
```

**Reply by thy960112** [reply]
Thank you very much for your quick reply, it's very useful and the `${KEYWORD}` and `${TAG_NAME}` in the last two lines may have to be replaced with `${KEYWORDS}` and `${KEYWORDS}`.

**Reply by yuzixieshouhuajiaqi** [reply]
大佬，英文文献只返回了添加标签成功[Conditional Tag] item ${item.getField("title")} with keyword ${KEYWORDS} is  tagged ${TAG_NAMES}，但是实际没有添加任何标签，不知道怎么回事，其他插件也都禁用了，只有中文文献有成功提示并成功添加标签。

**Reply by yuzixieshouhuajiaqi** [reply]
又试了几次，中文也不行了，只有添加标签成功的提示，实际没有添加标签

**Reply by zyx335588** [reply]
@windingwind I found when I used zotero connector to import an item, the tag was added at the time that the item information was imported, but when the pdf attachment was added or the attachment adding failed, the tag was removed again. Any help?

**Reply by windingwind** [reply]
@yuzixieshouhuajiaqi @zyx335588 fixed. please update the action data.

**Reply by yuzixieshouhuajiaqi** [reply]
@windingwind 大佬，这次运行结果正常了，可以添加标签了，但是每次运行都会有错误提示
XML Parsing Error: no root element found
Location: chrome://zotero/content/progressWindow.xhtml
Line Number 1, Column 1:

**Reply by windingwind** [reply]
检查升级插件版本

**Reply by yuzixieshouhuajiaqi** [reply]
大佬,是最新版1.0.0-15呀，还是有错误提示，不过结果都正常，也成功添加了

**Reply by windingwind** [reply]
那和插件也许无关，不管即可

**Reply by yuzixieshouhuajiaqi** [reply]
// Auto-tag when keywords detected in title
// @author windingwind
// @link https://github.com/windingwind/zotero-actions-tags/discussions/119
// @usage Replace the `KEYWORDS` and `TAG_NAMES`.
// @see https://github.com/windingwind/zotero-actions-tags/issues/77
// @see https://github.com/windingwind/zotero-actions-tags/issues/85

if (!item) {
    return "[Conditional Tag] item is empty";
}

const KEYWORDS1 = ["meta", "Meta"];
const TAG_NAMES1 = ["♏"];
const KEYWORDS2 = ["review", "Review", "现状", "进展"];
const TAG_NAMES2 = ["®"];
const KEYWORDS3 = ["case", "Case"];
const TAG_NAMES3 = ["©"];

if (KEYWORDS1.some(kwd => item.getField("title").includes(kwd))) {
    TAG_NAMES1.forEach(tag => item.addTag(tag, 1));
    return `[Conditional Tag] item ${item.getField("title")} with keyword ${KEYWORDS1} is tagged ${TAG_NAMES1}`;
} else if (KEYWORDS2.some(kwd => item.getField("title").includes(kwd))) {
    TAG_NAMES2.forEach(tag => item.addTag(tag, 1));
    return `[Conditional Tag] item ${item.getField("title")} with keyword ${KEYWORDS2} is tagged ${TAG_NAMES2}`;
} else if (KEYWORDS3.some(kwd => item.getField("title").includes(kwd))) {
    TAG_NAMES3.forEach(tag => item.addTag(tag, 1));
    return `[Conditional Tag] item ${item.getField("title")} with keyword ${KEYWORDS3} is tagged ${TAG_NAMES3}`;
} else {
    return `[Conditional Tag] item ${item.getField("title")} does not contain any of the specified keywords`;
}

**Reply by yuzixieshouhuajiaqi** [reply]
大佬，是我用了多种条件的原因嘛，这是gpt帮我改的，还是有那个错误提示

**Comment by huachuman** [comment]
I keep getting SyntaxError: return not in function?

**Reply by windingwind** [reply]
update the plugin, or please provide steps to reproduce.

**Reply by huachuman** [reply]
Here are the steps to reproduce:

Click the copy button in the code you provided

Paste copied text in "Run Javascript" window in Zotero.

Click Run.

**Reply by windingwind** [reply]
No. This is not for `run js`. See readme.

**Comment by BoBo-YunYun** [comment]
以下代码可以同时检测标题或摘要中是否含有某个字段，并返回相应标签：

```js
// Auto-tag based on specific keywords in the title and abstract
// @author windingwind
// @link https://github.com/windingwind/zotero-actions-tags/discussions/119
// @usage Replace the `KEYWORDS_AND_TAGS` array with your own keywords and corresponding tags.
// @see https://github.com/windingwind/zotero-actions-tags/issues/77
// @see https://github.com/windingwind/zotero-actions-tags/issues/85

if (!item) {
    return "[Conditional Tag] item is empty";
}

const KEYWORDS_AND_TAGS = [
    { keyword: "FIELD_A", tag: "TAG_B" },
    { keyword: "FIELD_C", tag: "TAG_D" },
    // You can add more keyword-tag pairs as needed
];

const title = item.getField("title");
const abstract = item.getField("abstract");
let tagged = false;

for (const pair of KEYWORDS_AND_TAGS) {
    const { keyword, tag } = pair;
    if (title.includes(keyword) || (abstract && abstract.includes(keyword))) {
        item.addTag(tag, 1);
        tagged = true;
    }
}

if (tagged) {
    return `[Conditional Tag] item "${title}" has been tagged based on keywords in title or abstract.`;
} else {
    return `[Conditional Tag] item "${title}" does not match any keywords in the list in title or abstract.`;
}
```

**Reply by windingwind** [reply]
脚本内容需用markdown格式
\`\`\`js
// code
\`\`\`

包裹。

**Reply by BoBo-YunYun** [reply]
好的

**Reply by fedroy** [reply]
Old version: `const abstract = item.getField("abstract");`
New version: `const abstract = item.getField("abstractNote");`

**Reply by Andrei-WongE** [reply]
I get an ``missing: after property id `` error message

**Reply by thy960112** [reply]
该版本可以实现新版本 Zotero 下

1. 同时检测标题或摘要中是否含有某个字段，并返回相应标签
2. 不区分字段大小写

```JavaScript
// Auto-tag based on specific keywords in the title and abstract
// @author windingwind
// @link https://github.com/windingwind/zotero-actions-tags/discussions/119
// @usage Replace the `KEYWORDS_AND_TAGS` array with your own keywords and corresponding tags.
// @see https://github.com/windingwind/zotero-actions-tags/issues/77
// @see https://github.com/windingwind/zotero-actions-tags/issues/85

if (!item) {
    return "[Conditional Tag] item is empty";
}

const KEYWORDS_AND_TAGS = [
	{ keyword: "survey", tag: "#Survey" },
	{ keyword: "hitchhiker", tag: "#Survey" },
    // You can add more keyword-tag pairs as needed
];


const title = item.getField("title").toLowerCase();
const abstract = item.getField("abstractNote").toLowerCase();
let tagged = false;

for (const pair of KEYWORDS_AND_TAGS) {
    const { keyword, tag } = pair;
    if (title.includes(keyword) || (abstract && abstract.includes(keyword))) {
        item.addTag(tag, 1);
        tagged = true;
    }
}

if (tagged) {
    return `[Conditional Tag] item "${title}" has been tagged based on keywords in title or abstract.`;
} else {
    return `[Conditional Tag] item "${title}" does not match any keywords in the list in title or abstract.`;
}
```

**Reply by Andrei-WongE** [reply]
I still get the same error, my Zotero version is 7.09. I have only modified this section: 
``
const KEYWORDS_AND_TAGS = [
{ keyword: "activty space", tag: "#activty space"},
{ keyword: "review", tag: "#Review"},
{ keyword: "segregation", tag: "#Segregation"},
{ keyword: "boundaries", "boundary", tag: "#boundaries"},
    // You can add more keyword-tag pairs as needed
];
``

**Reply by thy960112** [reply]
I know where the problem lies now, because all the content after `//` here is understood as annotation content, you can choose
1. Delete `//You can add more keyword tag pairs as needed`
2. Use line breaks as shown below
```JavaScript
const KEYWORDS_AND_TAGS = [ 
{ keyword: "activty space", tag: "#activty space"}, { keyword: "review", tag: "#Review"}, { keyword: "segregation", tag: "#Segregation"}, { keyword: "boundaries", "boundary", tag: "#boundaries"}, 
// You can add more keyword-tag pairs as needed 
];
```

**Reply by Andrei-WongE** [reply]
Thanks for the quick reply. Still getting the same error:
```

// Auto-tag based on specific keywords in the title and abstract
// @author windingwind
// @link https://github.com/windingwind/zotero-actions-tags/discussions/119
// @usage Replace the `KEYWORDS_AND_TAGS` array with your own keywords and corresponding tags.
// @see https://github.com/windingwind/zotero-actions-tags/issues/77
// @see https://github.com/windingwind/zotero-actions-tags/issues/85

if (!item) {
    return "[Conditional Tag] item is empty";
}

const KEYWORDS_AND_TAGS = [ 
{ keyword: "activty space", tag: "#activty space"}, { keyword: "review", tag: "#Review"}, { keyword: "segregation", tag: "#Segregation"}, { keyword: "boundaries", "boundary", tag: "#boundaries"}, 
];

const title = item.getField("title").toLowerCase();
const abstract = item.getField("abstractNote").toLowerCase();
let tagged = false;

for (const pair of KEYWORDS_AND_TAGS) {
    const { keyword, tag } = pair;
    if (title.includes(keyword) || (abstract && abstract.includes(keyword))) {
        item.addTag(tag, 1);
        tagged = true;
    }
}

if (tagged) {
    return `[Conditional Tag] item "${title}" has been tagged based on keywords in title or abstract.`;
} else {
    return `[Conditional Tag] item "${title}" does not match any keywords in the list in title or abstract.`;
}


```

**Reply by thy960112** [reply]
I copied your code and ran it, and there is indeed this issue
The problem is here `{keywords: "boundaries", "boundaries", tag: "# boundaries"}`
If you modify it to `{keyword:' boundaries', tag: '# boundaries'}`, there should be no error

**Comment by yuzixieshouhuajiaqi** [comment]
感谢大佬回答，不过运行后没有成功添加标签
// Auto-tag based on specific keywords in the title and abstract
// @author windingwind
// @link #119
// @Usage Replace the KEYWORDS_AND_TAGS array with your own keywords and corresponding tags.
// @see #77
// @see #85

if (!item) {
return "[Conditional Tag] item is empty";
}

const KEYWORDS_AND_TAGS = [
{ keyword: "meta", "Meta", tag: "♏" },
{ keyword: "review", "Review", "现状", "进展", "综述", tag: "®" },
{ keyword: "case", "Case", tag: "©" },
// You can add more keyword-tag pairs as needed
];

const title = item.getField("title");
const abstract = item.getField("abstract");
let tagged = false;

for (const pair of KEYWORDS_AND_TAGS) {
const { keyword, tag } = pair;
if (title.includes(keyword) || (abstract && abstract.includes(keyword))) {
item.addTag(tag, 1);
tagged = true;
}
}

if (tagged) {
return [Conditional Tag] item "${title}" has been tagged based on keywords in title or abstract.;
} else {
return [Conditional Tag] item "${title}" does not match any keywords in the list in title or abstract.;
}

**Reply by windingwind** [reply]
建议先用gpt做基础错误排除，或者了解一下js语法。

**Comment by huachuman** [comment]
> No. This is not for `run js`. See readme.

oh I see so this is also only for Z7

**Comment by RedIsND** [comment]
Hello,
I'm new to Zotero and to JavaScript.
I don't understand where I am supposed to run the script. I tried running it in developper tools - JavaScript, but it didn't work and I can't find where else I am supposed to run it. 
Thank you.

**Comment by Geo123abc** [comment]
我感觉是需要脚本维护一个  tags 库，我在想能不能这样：
1. 添加新条目时
2. 读取zotero中的所有tags
3. 逐个在标题或者摘要中搜索tags（可能需要设置大小写不敏感），如果搜到这个关键词，那么就添加对应的tag

**Reply by Geo123abc** [reply]
这个可能能达到目的 https://github.com/windingwind/zotero-actions-tags/discussions/268

**Comment by pekideer** [comment]
使用这个脚本添加了一堆tags，请问作者有没有批量删除它们的脚本。。。（好像zutilo有此功能）

**Reply by thy960112** [reply]
https://github.com/windingwind/zotero-actions-tags/discussions/127 这个可以删除条目的所有标签

**Reply by pekideer** [reply]
但是报错/(ㄒoㄒ)/~~

**Reply by thy960112** [reply]
我这边使用讨论最开始的如下一行代码，在条目上使用应该是没有的，评论报错似乎是在 PDF 阅读器页面上使用导致的。

`item.getTags().map((tag) => tag.tag).forEach((tag) => {item.removeTag(tag)})`

**Comment by thy960112** [comment]
Inspired by the code from [windingwind](https://github.com/windingwind/zotero-actions-tags/discussions/119#discussioncomment-7011649) and [BoBo-YunYun](https://github.com/windingwind/zotero-actions-tags/discussions/119#discussioncomment-7143933), I wrote the following script:

**Environment**
- Zotero 7.0.9  
- Zotero Actions and Tag 2.0.5  

**Features**
- Adds all corresponding tags from a predefined `tags` list when any keyword from the `keywords` list is detected in the title or abstract of an item. 检测到条目标题或摘要包含 `keywords` 列表里的任意一个关键字时，添加对应的 `tags` 标签列表中的所有标签。
- The matching process is **case-insensitive**. 不区分关键字的大小写。

**Goal**
- To allow a single action to add as many tags as possible while minimizing the effort required to write individual rules. 一个 Action 尽可能多地添加多个标签，并减少规则撰写。

```JS
// Auto-tag when keyword list detected in title and abstract
// @author windingwind, BoBo-YunYun, thy960112
// @link https://github.com/windingwind/zotero-actions-tags/discussions/119
// @usage Replace the `KEYWORD_GROUPS`

// Configuration: Keywords and corresponding tags
// Add more keyword-tag groups as needed
const KEYWORD_GROUPS = [
    {
        keywords: ["KWD1", "KWD2", "KWD3"],
        tags: ["TAG1", "TAG2"]
    },
    {
        keywords: ["NEW_KWD1", "NEW_KWD2"],
        tags: ["#NEW_TAG1"]
    }
];


if (!item) {
    return "[Conditional Tag] item is empty";
}

const title = item.getField("title").toLowerCase();
const abstract = item.getField("abstractNote").toLowerCase();
let tagsAdded = [];

// Process each keyword group
for (const group of KEYWORD_GROUPS) {
    // Check if any keyword from the group is found in title or abstract
    if (group.keywords.some(kwd => 
        title.includes(kwd.toLowerCase()) || 
        (abstract && abstract.includes(kwd.toLowerCase()))
    )) {
        // Add all corresponding tags for this keyword group
        group.tags.forEach(tag => {
            item.addTag(tag, 1);
            tagsAdded.push(tag);
        });
    }
}

// Generate return message
if (tagsAdded.length > 0) {
    const uniqueTags = [...new Set(tagsAdded)]; // Remove duplicate tags in message
    return `[Conditional Tag] item "${item.getField("title")}" has been tagged with: ${uniqueTags.join(", ")}`;
} else {
    return `[Conditional Tag] item "${item.getField("title")}" does not match any configured keywords`;
}
```

**Reply by Andrei-WongE** [reply]
@thy960112 as a curiosity, and take this from a non-expert, would it be possible to draw from the list of existing tags?

**Reply by thy960112** [reply]
This script may help you https://github.com/windingwind/zotero-actions-tags/discussions/268
